optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/entity/EntitySelectorFactory.java

```diff
@@ -0,0 +1,396 @@
+        SelectionCacheType resolvedCacheType = SelectionCacheType.resolve(config.getCacheType(), minimumCacheType);
+        SelectionOrder resolvedSelectionOrder = SelectionOrder.resolve(config.getSelectionOrder(), inheritedSelectionOrder);
+
+        if (config.getNearbySelectionConfig() != null) {
+            config.getNearbySelectionConfig().validateNearby(resolvedCacheType, resolvedSelectionOrder);
+        }
+        validateCacheTypeVersusSelectionOrder(resolvedCacheType, resolvedSelectionOrder);
+        validateSorting(resolvedSelectionOrder);
+        validateProbability(resolvedSelectionOrder);
+        validateSelectedLimit(minimumCacheType);
+
+        // baseEntitySelector and lower should be SelectionOrder.ORIGINAL if they are going to get cached completely
+        boolean baseRandomSelection = determineBaseRandomSelection(entityDescriptor, resolvedCacheType, resolvedSelectionOrder);
+        SelectionCacheType baseSelectionCacheType = SelectionCacheType.max(minimumCacheType, resolvedCacheType);
+        EntitySelector entitySelector = buildBaseEntitySelector(entityDescriptor, baseSelectionCacheType, baseRandomSelection);
+        if (config.getNearbySelectionConfig() != null) {
+            // TODO Static filtering (such as movableEntitySelectionFilter) should affect nearbySelection
+            entitySelector = applyNearbySelection(configPolicy, config.getNearbySelectionConfig(), minimumCacheType,
+                    resolvedSelectionOrder, entitySelector);
+        }
+        entitySelector = applyFiltering(entitySelector);
+        entitySelector = applySorting(resolvedCacheType, resolvedSelectionOrder, entitySelector);
+        entitySelector = applyProbability(resolvedCacheType, resolvedSelectionOrder, entitySelector);
+        entitySelector = applyShuffling(resolvedCacheType, resolvedSelectionOrder, entitySelector);
+        entitySelector = applyCaching(resolvedCacheType, resolvedSelectionOrder, entitySelector);
+        entitySelector = applySelectedLimit(resolvedSelectionOrder, entitySelector);
+        entitySelector = applyMimicRecording(configPolicy, entitySelector);
+        return entitySelector;
+    }
+
+    protected EntitySelector buildMimicReplaying(HeuristicConfigPolicy configPolicy) {
+        if (config.getId() != null
+                || config.getEntityClass() != null
+                || config.getCacheType() != null
+                || config.getSelectionOrder() != null
+                || config.getNearbySelectionConfig() != null
+                || config.getFilterClass() != null
+                || config.getSorterManner() != null
+                || config.getSorterComparatorClass() != null
+                || config.getSorterWeightFactoryClass() != null
+                || config.getSorterOrder() != null
+                || config.getSorterClass() != null
+                || config.getProbabilityWeightFactoryClass() != null
+                || config.getSelectedCountLimit() != null) {
```

Comment:
Good that you're not doing mutation testing. You'd grow old testing all the permutations of this particular condition.

Reply:
But to be constructive here - I'd add all of the calls into a stream, and then filter anything non null. That way, the code would  more readily express what you're trying to do.

Reply:
Good point, I used the same "trick" for resolving terminationConfigs; here it suits pretty well too.